lightning/src/ln/script.rs

```diff
@@ -0,0 +1,285 @@
+	/// serialization.
+	Legacy(PublicKey),
+
+	/// [`Script`] adhering to a script pubkey format specified in BOLT #2.
+	Bolt2(Script),
+}
+
+impl Writeable for ShutdownScript {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), ::std::io::Error> {
+		self.0.write(w)
+	}
+
+	fn serialized_length(&self) -> usize {
+		self.0.serialized_length()
+	}
+}
+
+impl Readable for ShutdownScript {
+	fn read<R: Read>(r: &mut R) -> Result<Self, DecodeError> {
+		Ok(ShutdownScript(ShutdownScriptImpl::read(r)?))
+	}
+}
+
+impl_writeable_tlv_based_enum!(ShutdownScriptImpl, ;
+	(0, Legacy),
+	(1, Bolt2),
+);
+
+impl ShutdownScript {
+	/// Generates a P2PKH script pubkey from the given [`PubkeyHash`].
+	pub fn new_p2pkh(pubkey_hash: &PubkeyHash) -> Self {
+		Self(ShutdownScriptImpl::Bolt2(Script::new_p2pkh(pubkey_hash)))
+	}
+
+	/// Generates a P2SH script pubkey from the given [`ScriptHash`].
+	pub fn new_p2sh(script_hash: &ScriptHash) -> Self {
+		Self(ShutdownScriptImpl::Bolt2(Script::new_p2sh(script_hash)))
+	}
+
+	/// Generates a P2WPKH script pubkey from the given [`WPubkeyHash`].
+	pub fn new_p2wpkh(pubkey_hash: &WPubkeyHash) -> Self {
+		Self(ShutdownScriptImpl::Bolt2(Script::new_v0_wpkh(pubkey_hash)))
+	}
+
+	/// Generates a P2WSH script pubkey from the given [`WScriptHash`].
+	pub fn new_p2wsh(script_hash: &WScriptHash) -> Self {
+		Self(ShutdownScriptImpl::Bolt2(Script::new_v0_wsh(script_hash)))
+	}
+
+	/// Generates a P2WSH script pubkey from the given segwit version and program.
+	///
+	/// # Panics
+	///
+	/// This function may panic if given a segwit program with an invalid length.
+	pub fn new_witness_program(version: u5, program: &[u8]) -> Self {
+		let script = Script::new_witness_program(version, program);
+		Self::try_from(script).expect("Invalid segwit program")
+	}
+
+	/// Converts the shutdown script into the underlying [`Script`].
+	pub fn into_inner(self) -> Script {
+		self.into()
+	}
+
+	/// Returns the [`PublicKey`] used for a P2WPKH shutdown script if constructed directly from it.
+	pub fn as_legacy_pubkey(&self) -> Option<&PublicKey> {
+		match &self.0 {
+			ShutdownScriptImpl::Legacy(pubkey) => Some(pubkey),
+			ShutdownScriptImpl::Bolt2(_) => None,
+		}
+	}
+
+	/// Returns whether the shutdown script is compatible with the features as defined by BOLT #2.
+	///
+	/// Specifically, checks for compliance with feature `option_shutdown_anysegwit`.
+	pub fn is_compatible(&self, features: &InitFeatures) -> bool {
+		match &self.0 {
+			ShutdownScriptImpl::Legacy(_) => true,
+			ShutdownScriptImpl::Bolt2(script) => is_bolt2_compliant(script, features),
+		}
+	}
+}
+
+fn is_bolt2_compliant(script: &Script, features: &InitFeatures) -> bool {
+	if script.is_p2pkh() || script.is_p2sh() || script.is_v0_p2wpkh() || script.is_v0_p2wsh() {
+		true
+	} else if features.supports_shutdown_anysegwit() {
+		script.is_witness_program() && script.as_bytes()[0] != SEGWIT_V0.into_u8()
```

Comment:
It looks like the `script.len() > 42` check from `is_unsupported_shutdown_script` didn't make it to here?

Reply:
So that check is redundant with `script.is_witness_program()`, although keeping it would prevent copying any absurdly large scripts from peers. Since I'm using `TryFrom` now, it would need to be performed at the call sites -- `clone` is used there because the script comes from inside the message sent by the peer which is passed by reference. For our own scripts, no conversion is necessary since `KeysInterface` is defined in terms of `ShutdownScript`.

That said, I'm not sure if the optimization / DoS concerns are warranted given the limitations on message sizes. May make sense to pass the message by value if we want to avoid clones, assuming we can destruct it while placating the compiler.

Reply:
> So that check is redundant with script.is_witness_program(), although keeping it would prevent copying any absurdly large scripts from peers. 

Ah! I hadn't realized that. I think in the old code it made me feel better cause it made it explicit, but I don't feel strongly.